src/test/java/com/google/sps/MatchRankingTest.java

```diff
@@ -361,18 +366,53 @@ public void testDifferentPercentKeywordBios() {
     User userBad = new User("");
     userBad.setId("123456");
 
-    //test different percentages of keywords against a 0% keyword bio
-    for(double i = 0.05; i <= 1.0; i += 0.05) {
-      userGood.setBio(generateRandomBio(50, i));
-      userBad.setBio(generateRandomBio(50, 0.0));
+    //put minimum benchmark checks in map (maps keyword percent to percent it should be ranked first)
+    Map<Integer, Double> expectedPerformance = new HashMap<Integer, Double>();
+    expectedPerformance.put(0, 0.0);
+    expectedPerformance.put(25, 0.0);
+    expectedPerformance.put(50, 0.2);
```

Comment:
These values are fine (no action requested); Out of curiosity, were these based off the output - some constant, or did you use some theory/empirical evaluation to pick values that would consistently pass the test?

If you're interested / for future reference, a more concrete way of figuring out how much lower than the expected value you should use is to use the properties of the binomial:

Imagine you run one of your cases 1000 times (in order to calibrate), and find that it passes (has the right ranking order) 70% of the time. Then, suppose that, in the test (not just for calibration) you perform 100 repetitions; On average you'd expect to see 70 of them pass, however if you look at the binomial distribution, you'll see that there's still a ~2% probability that you might get as few as 60 passing (here's a [binomial calculator](https://www.wolframalpha.com/widgets/view.jsp?id=78baf4f3a070cc5b9b226664d2ce80ec), I got the 2% by plugging in n=100,  p=0.7, X>60, and then subtracting 100% - 0.97011 (calculator result)) . 

You can then define an acceptable failure rate and use that to define the lower bound (e.g.: perhaps you're ok with the test failing 1:100000 times / a 99.99% success rate -- "four nines" as we'd call it), which in the above example works out to 51 (that is, in 1:100000 cases the ranking order will be correct less than 51 times, causing the test to fail if we use this as the threshold).

That's one way of dealing with pseudo-random tests. In general though, this kind of statistical reasoning is hidden away by the benchmarking tool, and in regular tests we mostly try to avoid making tests random exactly because properly testing semi-random behavior is hard (though sometimes inevitable / better than the alternatives, like here where we don't have real data to work with).

Reply:
I got the values by running the test a few times then picking percents a little bit under what I got for the output to make sure it would almost always pass. That's an interesting way of calculating it - I did find that if I ran the test a bunch of times it would occasionally get a really low percentage and fail, so I tried to make the benchmarks low enough that this wouldn't happen too often :)